Fixes output_token not in output_token.names for call_torchbind operator

ydwu4 opened the pull request:
Differential Revision: D58767610

We're getting an assertion error for `output_token not in output_token_names`. 

The error is pretty tricky:
1. in replace_set_grad_with_hop pass, the graph_signature doesn't change correctly according to the graph change of the pass. Related changes: replace_set_grad_with_hop_pass.py, graph_signature.py
2. _remove_effect_tokens pass collect the same call_torchbind operator twice. TODO is to support remove_effect_tokens in subgraphs. Related changes: _remove_effect_tokens_pass.py

angelayi commented on torch/export/_remove_effect_tokens_pass.py:
```diff
@@ -44,6 +44,7 @@ def _remove_effect_tokens_from_graph_helper(
     output_node.args = (tuple(output_args[num_tokens:]),)
     for out_token in out_token_nodes:
         assert out_token.name in output_token_names
```
can you help me add an assertion message here so that (hopefully) it is easier to debug in the future?

angelayi commented on torch/_export/utils.py:
```diff
@@ -300,9 +300,9 @@ def get_lifted_tensor_constant(
-    Splits the graph module into multiple submodules based on the node_call_back.
-    The node_call_back should return True if the node is a delimiter. Delimiter will be
-    the first node in the next submodule.
+    sequential_split creates a new graph module that splits the input graph module into multiple submodules 
```
👌💯💪

angelayi commented on torch/_export/passes/replace_set_grad_with_hop_pass.py:
```diff
@@ -147,11 +147,21 @@ def _sequential_split_and_maybe_inline_subgraphs(
+            new_signature = copy.deepcopy(graph_signature)
+            new_gm_out_node =  next(reversed(new_gm.graph.find_nodes(op="output")))
+            assert new_gm_out_node.op == "output" and len(new_gm_out_node.args[0]) == len(new_signature.output_specs)
+            for arg_node, out_spec in zip(new_gm_out_node.args[0], new_signature.output_specs):
+                if out_spec.arg.name != arg_node.name:
+                    out_spec.arg.name = arg_node.name
```
I'm surprised _set_replace_hook doesn't already do this?

ydwu4 replied:
Yeah, but it can only do its job when the graph module being inplace transformed and graph signature has a correct correspondence.

The reason we need to this mannual mapping is that, sequential_split createa a new graph module that has inconsistent output names with the name stored in graph signature.output_specs. As a result, if we don't mannually override the output names, even if we set the new_graph_module's replace hook to be new_signature's replace_hook, the replace_hook function cannot find the original name in the output_specs cauing the replacement fail silently [here](https://github.com/pytorch/pytorch/pull/129146/files#diff-92420f977c3a02b2deadf6752ce4a9ee601c20612a1a13cc365252eb09410edbL516). One possible better way is we should error out when we cannot find match any old names.

angelayi replied:
yeah that makes sense! thanks for the explanation! 
> One better way is we should error out when we cannot find match any old names in the replace hook to avoid this kind of silent incorrectness.
I'm not sure if erroring will cause any issues.. can do in a separate PR I guess :P

ydwu4 committed:
Fixes output_token not in output_token.names for call_torchbind operator (#129146)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/129146

We're getting an assertion error for `output_token not in output_token_names`.

The error is pretty tricky:
1. in replace_set_grad_with_hop pass, the graph_signature doesn't change correctly according to the graph change of the pass. Related changes: replace_set_grad_with_hop_pass.py, graph_signature.py
2. _remove_effect_tokens pass collect the same call_torchbind operator twice. TODO is to support remove_effect_tokens in subgraphs. Related changes: _remove_effect_tokens_pass.py

Test Plan: Tests in T192448049.

Reviewed By: angelayi

Differential Revision: D58767610

huydhn commented:
@pytorchbot merge

kit1980 commented:
@ydwu4 please give more details in the title a description. "Fixes T192448049" is completely meaningless for open source users and developers.

ydwu4 commented:
> @ydwu4 please give more details in the title and description.

Done. Thanks for pointing it out. I was expecting when i re-export the internal diff, the OSS one will also sync the title and description.
